Web Inspector: Layers tab should show a visual representation of the layer

AnujPanta1 opened the pull request:
#### 79db54bf56d12e55ab0e60eb1e7db6827a9d0f58
<pre>
Web Inspector: Layers tab should show a visual representation of the layer
<a href="https://bugs.webkit.org/show_bug.cgi?id=302723">https://bugs.webkit.org/show_bug.cgi?id=302723</a>
<a href="https://rdar.apple.com/problem/164977594">rdar://problem/164977594</a>

Reviewed by Devin Rousso.

Implemented layer snapshot rendering to display actual composited content in the 3D Layers view.
Added LayerTree.snapshotForLayer protocol command that uses
GraphicsLayer::paintGraphicsLayerContents to capture rendered output as PNG data URLs. The
frontend loads these snapshots as Three.js textures and maps them onto layer meshes with proper
UV coordinates, replacing the previous solid-color placeholder rectangles. This approach provides
direct visualization of what each compositing layer actually renders, making it easier to debug
compositing issues and understand layer content at a glance.

* Source/JavaScriptCore/inspector/protocol/LayerTree.json:
* Source/WebCore/inspector/agents/InspectorLayerTreeAgent.cpp:
(WebCore::InspectorLayerTreeAgent::requestContent):
* Source/WebCore/inspector/agents/InspectorLayerTreeAgent.h:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Controllers/LayerTreeManager.js:
(WI.LayerTreeManager.prototype.snapshotForLayer):
* Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
(WI.Layers3DContentView):
(WI.Layers3DContentView.prototype.get navigationItems):
(WI.Layers3DContentView.prototype.attached):
(WI.Layers3DContentView.prototype.detached):
(WI.Layers3DContentView.prototype._updateLayers):
(WI.Layers3DContentView.prototype._createLayerGroup):
(WI.Layers3DContentView.prototype._populateLayerGroup):
(WI.Layers3DContentView.prototype._loadLayerTexture):
(WI.Layers3DContentView.prototype._createLayerMesh):
(WI.Layers3DContentView.prototype._handleLayerContentsSettingChanged):
(WI.Layers3DContentView.prototype._handleRefreshLayersButtonClicked):
(WI.Layers3DContentView.prototype._refreshAllLayers):
(WI.Layers3DContentView.prototype._zInterval):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createExperimentalSettingsView):

Canonical link: <a href="https://commits.webkit.org/304758@main">https://commits.webkit.org/304758@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd797ddbf0d497561b9708822974651cb4a466de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133552 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44710 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141112 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85604 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/715737b5-0b47-4422-b091-4aad48f4654e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135422 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6577 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5921 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102171 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69539 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e414bc92-ebeb-4d2e-a826-2f990643ddc5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4661 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119737 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82967 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e42ca88d-a9ef-44f8-9bcf-5a2d4703d33f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4539 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2151 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/498 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125629 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113674 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37851 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143761 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132066 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5729 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38431 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110546 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5809 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110723 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28068 "") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4395 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115989 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59478 "The change is no longer eligible for processing.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21037 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5781 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34304 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165029 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5628 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69239 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43124 "Found 327 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Regex/captures.js.default, ChakraCore.yaml/ChakraCore/test/Regex/regex1.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/NoBacktrackingChomp.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/acid.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/assertion.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/captures.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/class-case.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/es5SpecExamples.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/fastRegexCaptures.js.default, ChakraCore.yaml/ChakraCore/test/UnifiedRegex/match_global.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5872 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5736 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/protocol/LayerTree.json:
```diff
@@ -108,6 +108,16 @@
             ]
+        },
+        {
+            "name": "snapshotForLayer",
```
i think to match similar commands in other domains this might be better named as just `requestContent`

dcrousso commented on Source/JavaScriptCore/inspector/protocol/LayerTree.json:
```diff
@@ -108,6 +108,16 @@
+                { "name": "layerId", "$ref": "LayerId", "description": "The id of the layer to snapshot." }
+            ],
+            "returns": [
+                { "name": "dataURL", "type": "string", "description": "Base64-encoded PNG data URL of the layer's rendered content." }
```
along those lines id also just call this `content`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/LayerTreeManager.js:
```diff
@@ -197,6 +197,21 @@ WI.LayerTreeManager = class LayerTreeManager extends WI.Object
+        console.assert(this.supported);
+
+        let target = WI.assumingMainTarget();
+        target.LayerTreeAgent.snapshotForLayer(layer.layerId, 2.0, function(error, dataURL) {
```
what's the `2.0` for?

AnujPanta1 replied:
this was left in by mistake, I was once trying to implement a resolution parameter but went through

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -41,6 +41,18 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
 
+        this._showLayerContentButtonNavigationItem = new WI.ActivateButtonNavigationItem("show-layer-content", WI.UIString("Show layer content"), WI.UIString("Hide layer content"), "Images/Camera.svg", 16, 16);
+        this._showLayerContentButtonNavigationItem.addEventListener(WI.ButtonNavigationItem.Event.Clicked, this._handleShowLayerContentButtonClicked, this);
+        this._showLayerContentButtonNavigationItem.activated = true;
```
we should make this into a `WI.Setting` so that if the developer wants to keep it off then Web Inspector respects that instead of always defaulting back to on

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -41,6 +41,18 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        this._refreshLayersButtonNavigationItem.visibilityPriority = WI.NavigationItem.VisibilityPriority.Low;
+
+        this._showLayerContent = true;
+        this._pendingTextureLoads = new Map();
```
Style: we don't add `()` when not providing arguments with `new`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -41,6 +41,18 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        this._showLayerContentButtonNavigationItem.activated = true;
+        this._showLayerContentButtonNavigationItem.visibilityPriority = WI.NavigationItem.VisibilityPriority.Low;
+
+        this._refreshLayersButtonNavigationItem = new WI.ButtonNavigationItem("refresh-layers", WI.UIString("Refresh layer snapshots"), "Images/ReloadToolbar.svg", 15, 15);
```
i feel like we should at least make this `disabled` when `!this._showLayerContentButtonNavigationItem.activated` but perhaps more ideally not even show this navigation item in that case

dcrousso replied:
i actually even wonder whether a developer would really even turn off layer snapshots as it seems like it's strictly significantly more useful to at least have _some_ content in each layer (especially if new layers automatically grab contents and there's this way of refreshing all layers)

as such, perhaps we should make `_showLayerContentButtonNavigationItem` into a checkbox in the Settings Tab instead?  i really doubt that developers will frequently turn this on/off so moving it out of the way (while still providing the choice) seems like a nice middle ground

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -41,6 +41,18 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
         this._paintFlashingButtonNavigationItem.enabled = WI.LayerTreeManager.supportsShowingPaintRects();
         this._paintFlashingButtonNavigationItem.visibilityPriority = WI.NavigationItem.VisibilityPriority.Low;
 
+        this._showLayerContentButtonNavigationItem = new WI.ActivateButtonNavigationItem("show-layer-content", WI.UIString("Show layer content"), WI.UIString("Hide layer content"), "Images/Camera.svg", 16, 16);
```
im not sure the camera is the best choice here, but i cant think of a better option so i think it's fine for now

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -41,6 +41,18 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        this._showLayerContentButtonNavigationItem.activated = true;
+        this._showLayerContentButtonNavigationItem.visibilityPriority = WI.NavigationItem.VisibilityPriority.Low;
+
+        this._refreshLayersButtonNavigationItem = new WI.ButtonNavigationItem("refresh-layers", WI.UIString("Refresh layer snapshots"), "Images/ReloadToolbar.svg", 15, 15);
```
you use the term "layer content" for the other navigation item so i think we should have it be "Refresh layer contents" here to match

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -41,6 +41,18 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        this._refreshLayersButtonNavigationItem.addEventListener(WI.ButtonNavigationItem.Event.Clicked, this._handleRefreshLayersButtonClicked, this);
+        this._refreshLayersButtonNavigationItem.visibilityPriority = WI.NavigationItem.VisibilityPriority.Low;
+
+        this._showLayerContent = true;
```
we should move this up and use it in the `.activated =` instead of hardcoding `true` twice

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -66,7 +78,7 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
     get navigationItems()
     {
-        return [this._compositingBordersButtonNavigationItem, this._paintFlashingButtonNavigationItem];
+        return [this._compositingBordersButtonNavigationItem, this._paintFlashingButtonNavigationItem, this._showLayerContentButtonNavigationItem, this._refreshLayersButtonNavigationItem];
```
Style: lets break this up onto multiple lines for readability

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +296,88 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        });
+    }
+
+    _createLayerMesh({x, y, width, height}, isOutline = false, texture = null)
```
Style: we prefer options objects instead of parameters with default values
```js
_createLayerMesh({x, y, width, height}, {isOutline, texture} = {})
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +296,88 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
-        layerGroup.add(this._createLayerMesh(layer.bounds), this._createLayerMesh(layer.compositedBounds, true));
+
+        let fillMesh = this._createLayerMesh(layer.bounds, false, null);
+        let outlineMesh = this._createLayerMesh(layer.compositedBounds, true, null);
```
no need to provide `false` or `null` since those are the default values

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +296,88 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+                    return;
+                }
+
+                layerGroup.remove(fillMesh);
```
rather than removing each of these individually, should we just call `layerGroup.clear()` and then add back the ones we want?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +296,88 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+                    if (fillMesh.material.map)
+                        fillMesh.material.map.dispose();
+                    fillMesh.material.dispose();
+                }
```
we can simplify this with optional chaining
```js
fillMesh.geometry?.dispose();
fillMesh.material?.map?.dispose();
fillMesh.material?.dispose();
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +296,88 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
-    _createLayerMesh({x, y, width, height}, isOutline = false)
+    _loadLayerTexture(layer, callback) {
+        if (this._pendingTextureLoads.has(layer.layerId))
+            this._pendingTextureLoads.get(layer.layerId).cancelled = true;
```
this does two lookups so lets save to a temporary instead
```js
let pendingTextureLoad = this._pendingTextureLoads.get(layer.layerId);
if (pendingTextureLoad)
    pendingTextureLoad.cancelled = true;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +296,88 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+                return;
+            }
+
+            const textureLoader = new THREE.TextureLoader();
```
Style: please use `let` for anything that is not an actual constant
Style: we dont add `()` when no arguments are given with `new`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +296,88 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+            }
+
+            const textureLoader = new THREE.TextureLoader();
+            textureLoader.load(
```
Style: we don't really ever split function calls across multiple lines like this, so id suggest creating separate arguments instead
```js
let onLoad = (texture) => {
    /* ... */
};
const onProgress = undefined;
let onError = (error) => {
    /* ... */
};
textureLoader.load(content, onLoad, onProgress, onError);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -305,13 +394,30 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        if (texture) {
+            geometry.faceVertexUvs[0] = [
+                [new THREE.Vector2(0, 1), new THREE.Vector2(0, 0), new THREE.Vector2(1, 1)],
+                [new THREE.Vector2(0, 0), new THREE.Vector2(1, 0), new THREE.Vector2(1, 1)]
```
Style: missing trailing `,`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -417,6 +523,52 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+                if (child.material) {
+                    if (child.material.map)
+                        child.material.map.dispose();
+                    child.material.dispose();
```
ditto with optional chaining

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -417,6 +523,52 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        }
+        this._layerGroupsById.clear();
+
+        const zInterval = 25;
```
we have this `const` elsewhere, so we can we combine them into a `WI.Layers3DContentVIew._zInterval` so that they are both modified if the value is ever changed?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -417,6 +523,52 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+
+        const zInterval = 25;
+        this._layers.forEach((layer, index) => {
+            let layerGroup = this._createLayerGroup(layer);
```
is it really necessary to completely destroy and recreate every single group when refreshing the content of all layers?  could we instead just call `_loadLayerTexture` for each of these to just recreate the texture for the content?

AnujPanta1 replied:
yes great point, we now keep layers alive and call `loadLayerTexture` to avoid expensive cycles

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +296,88 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        if (this._pendingTextureLoads.has(layer.layerId))
+            this._pendingTextureLoads.get(layer.layerId).cancelled = true;
+
+        let loadContext = { cancelled: false };
```
it seems like the only reason we're using an object here is so we still have something to reference even if it's removed from `_pendingTextureLoads` (i.e. there's no other properties on this object other than `cancelled`).  this definitely doesn't feel like the most ideal approach, but im not sure there would be anything more efficient (other than maybe an incrementing "generation"-style value but that has it's own issues) so i suppose this is fine for now

AnujPanta1 replied:
Yeah it felt not ideal but the most practical solution

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +296,88 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        let loadContext = { cancelled: false };
+        this._pendingTextureLoads.set(layer.layerId, loadContext);
+
+        WI.layerTreeManager.snapshotForLayer(layer, (dataURL) => {
```
for complex pages this will result in a _lot_ of protocol traffic happening at once

i wonder if we should add a `requestContent` parameter to `LayerTree.layersForNode` that populates a new `content` property on the `LayerTree.Layer` object?

though this could mean that the response for `LayerTree.layersForNode` is gigantic due to having so much base64 data in it

id be curious to know which approach is faster and/or more efficient

AnujPanta1 replied:
I'm going to write this down as an investigation follow up, I did try across websites I thought would have many layers (like in the video I was on Airbnb) and didn't see any performance issues, however a deeper investigation would be useful I'll note it down for the future

dcrousso reviewed (commented):
this is really awesome!  great work :)

mostly just Style/NITs but a few functional suggestions as well

burg commented on Source/WebCore/inspector/agents/InspectorLayerTreeAgent.cpp:
```diff
@@ -338,6 +342,43 @@ Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::LayerTree::Compositi
     return compositingReasons;
 }
 
+Inspector::Protocol::ErrorStringOr<String> InspectorLayerTreeAgent::snapshotForLayer(const Inspector::Protocol::LayerTree::LayerId& layerId)
```
Nit: you can use the alias `CommandResult<void>` as the result type.

AnujPanta1 replied:
yes, using `CommandResult<string>` function returns string

burg commented on Source/WebCore/inspector/agents/InspectorLayerTreeAgent.cpp:
```diff
@@ -338,6 +342,43 @@ Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::LayerTree::Compositi
 
+Inspector::Protocol::ErrorStringOr<String> InspectorLayerTreeAgent::snapshotForLayer(const Inspector::Protocol::LayerTree::LayerId& layerId)
+{
+    const RenderLayer* renderLayer = m_idToLayer.get(layerId);
```
Nit: i think these can be `auto*`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/LayerTreeManager.js:
```diff
@@ -197,6 +197,21 @@ WI.LayerTreeManager = class LayerTreeManager extends WI.Object
+        let target = WI.assumingMainTarget();
+        target.LayerTreeAgent.requestContent(layer.layerId, function(error, content) {
+            if (error) {
+                console.error("Failed to get layer snapshot:", error);
```
lets use `WI.reportInternalError` so that it actually shows an error in local builds (i.e. more likely that someone sees it and reports it and we can fix it)

AnujPanta1 replied:
great call !

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -41,6 +41,13 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
 
+        this._refreshLayersButtonNavigationItem = new WI.ButtonNavigationItem("refresh-layers", WI.UIString("Refresh layer contents"), "Images/ReloadToolbar.svg", 15, 15);
+        this._refreshLayersButtonNavigationItem.addEventListener(WI.ButtonNavigationItem.Event.Clicked, this._handleRefreshLayersButtonClicked, this);
+        this._refreshLayersButtonNavigationItem.enabled = WI.settings.layers3DShowLayerContents.value;
```
normally id say that this shouldn't even be shown when `WI.settings.layers3DShowLayerContents` is disabled, but since we're having it be on-by-default i think it's OK for it to be toggling `.enabled` instead

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -41,6 +41,13 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
         this._paintFlashingButtonNavigationItem.enabled = WI.LayerTreeManager.supportsShowingPaintRects();
         this._paintFlashingButtonNavigationItem.visibilityPriority = WI.NavigationItem.VisibilityPriority.Low;
 
+        this._refreshLayersButtonNavigationItem = new WI.ButtonNavigationItem("refresh-layers", WI.UIString("Refresh layer contents"), "Images/ReloadToolbar.svg", 15, 15);
```
NIT: i wonder if just "Refresh layers" is enough?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -284,11 +298,81 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+        if (pendingTextureLoad)
+            pendingTextureLoad.cancelled = true;
+
+        let loadContext = { cancelled: false };
```
instead of creating an object that just has a `cancelled` property on it, perhaps we could use a uniquely created `Symbol` each time to check if the value in `_pendingTextureLoads` still matches this "context"?
```js
let symbol = Symbol("loading");
this._pendingTextureLoads.set(layer.layerId, symbol);

WI.layerTreeManager.snapshotForLayer(layer, (content) => {
    if (this._pendingTextureLoads.get(layer.layerId) !== symbol) {
        this._pendingTextureLoads.delete(layer.layerId);
        return;
    }

    // ...
});
```
it does require doing more `Map` lookups, but this way we dont create as many of those odd objects

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -417,6 +518,64 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
+            }
+            layerGroup.clear();
+
+            let fillMesh = this._createLayerMesh(layer.bounds);
```
i think the rest of this function is the same as what's done in `_createLayerGroup`.  could we create a separate function like `_populateLayerGroup` to avoid the repeated code?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
```diff
@@ -300,6 +300,13 @@ WI.SettingsTabContentView = class SettingsTabContentView extends WI.TabContentVi
             elementsSettingsView.addSeparator();
         }
 
+        if (InspectorBackend.hasDomain("LayerTree")) {
```
this should only really be when `LayerTree.requestContent` since older backends wont support that command

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
```diff
@@ -300,6 +300,13 @@ WI.SettingsTabContentView = class SettingsTabContentView extends WI.TabContentVi
 
+        if (InspectorBackend.hasDomain("LayerTree")) {
+            let layersGroup = elementsSettingsView.addGroup(WI.UIString("Layers:"));
+            layersGroup.addSetting(WI.settings.layers3DShowLayerContents, WI.UIString("Show layer contents"));
```
for now id probably make this a `WI.ExperimentalSetting` since it is somewhat new (see <https://github.com/WebKit/WebKit/pull/54125#discussion_r2540228762> about potential performance concerns)

eventually we'd wanna put this in a separate "Layers" pane since it has nothing to do with the Elements Tab

AnujPanta1 replied:
done, added to experimental

dcrousso commented on Source/WebCore/inspector/agents/InspectorLayerTreeAgent.cpp:
```diff
@@ -338,6 +342,43 @@ Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::LayerTree::Compositi
+    constexpr float scaleFactor = 2.0;
+    IntSize integralSize = IntSize(layerSize);
+
+    auto imageBuffer = ImageBuffer::create(integralSize, RenderingMode::Unaccelerated, RenderingPurpose::Snapshot, scaleFactor, DestinationColorSpace::SRGB(), PixelFormat::BGRA8);
```
what about other color spaces?

AnujPanta1 replied:
yeah, we could implement it I didn't because canvas and snapshots just used sRBG, but I can see how the in layers tab people might want more accuracy, I think somewhere in CanvasBase we get the colorSpace so ill do similar (perhaps filing a bug to update those other parts of WI to use colorSpace too might be worth it)

AnujPanta1 replied:
actually I think there are some fixe me's waiting around PageColorSampler ~ line 120 that are saying ""support sampling non-RGB values like p3" so I think ill leave sRBG for now, as a debugging tool I think it'll be okay

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
```diff
@@ -532,6 +691,7 @@ WI.Layers3DContentView = class Layers3DContentView extends WI.ContentView
     }
 };
 
+WI.Layers3DContentView._zInterval = 25;
```
looking at your demo video, i actually wonder if maybe we should lower this when the content is shown so that it's slightly easier to view the "flattened" page rendering

can definitely be a followup as i dont wanna derail this overall awesome feature :)

AnujPanta1 replied:
I added a somewhat dynamic version of zInterval but I have another patch coming that has to do with how we do the zInterval stuff so it will be used much more in that patch

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/LayerTreeManager.js:
diagnostic logging is something completely different from an internal error (i.e. it's about usage stats/analytics)

we should just use `WI.reportInternalError(error);` here instead like we do for a lot of other protocol errors

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
can we save `let zInterval = this._zInterval();` outside so that we dont have to check `WI.settings.experimentalLayers3DShowLayerContents` over and over in a loop?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
what was wrong with `layerGroup.clear()`?  if there's a reason why we can't use that then id suggest adding a brief comment here explaining why

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
i actually don't think we need to do anything here since you're about to override the value with a new `set` on the next line anyways, so id drop all of this

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
i think this is technically incorrect as if there's multiple requests in-flight then it would cancel the latest one, which is ideal since we would actually want the last request to (eventually) succeed

as such, i think we should only be calling `delete` when actually successful in `onLoad` and `onError`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
id somewhat prefer a `for (let i = 0; i < this._layers.length; ++i)` instead of having to invoke a function on each iteration, but if you prefer this then that's fine

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
ditto about moving this outside

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
ditto about why we can't use `layerGroup.clear()` here?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
now that we have `_zInterval` we don't need this anymore, right?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
we dont need this now that it's in the Experimental pane

dcrousso reviewed (approved):
nice work!  a few last things to fix, but overall i think this looks really good!  thanks for iterating with me :)

AnujPanta1 committed:
Web Inspector: Layers tab should show a visual representation of the layer
https://bugs.webkit.org/show_bug.cgi?id=302723
rdar://problem/164977594

Reviewed by Devin Rousso.

Implemented layer snapshot rendering to display actual composited content in the 3D Layers view.
Added LayerTree.snapshotForLayer protocol command that uses
GraphicsLayer::paintGraphicsLayerContents to capture rendered output as PNG data URLs. The
frontend loads these snapshots as Three.js textures and maps them onto layer meshes with proper
UV coordinates, replacing the previous solid-color placeholder rectangles. This approach provides
direct visualization of what each compositing layer actually renders, making it easier to debug
compositing issues and understand layer content at a glance.

* Source/JavaScriptCore/inspector/protocol/LayerTree.json:
* Source/WebCore/inspector/agents/InspectorLayerTreeAgent.cpp:
(WebCore::InspectorLayerTreeAgent::requestContent):
* Source/WebCore/inspector/agents/InspectorLayerTreeAgent.h:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Controllers/LayerTreeManager.js:
(WI.LayerTreeManager.prototype.snapshotForLayer):
* Source/WebInspectorUI/UserInterface/Views/Layers3DContentView.js:
(WI.Layers3DContentView):
(WI.Layers3DContentView.prototype.get navigationItems):
(WI.Layers3DContentView.prototype.attached):
(WI.Layers3DContentView.prototype.detached):
(WI.Layers3DContentView.prototype._updateLayers):
(WI.Layers3DContentView.prototype._createLayerGroup):
(WI.Layers3DContentView.prototype._populateLayerGroup):
(WI.Layers3DContentView.prototype._loadLayerTexture):
(WI.Layers3DContentView.prototype._createLayerMesh):
(WI.Layers3DContentView.prototype._handleLayerContentsSettingChanged):
(WI.Layers3DContentView.prototype._handleRefreshLayersButtonClicked):
(WI.Layers3DContentView.prototype._refreshAllLayers):
(WI.Layers3DContentView.prototype._zInterval):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createExperimentalSettingsView):

Canonical link: https://commits.webkit.org/304758@main
